csrc/ir/nodes.cpp

```diff
@@ -2111,7 +2111,14 @@ std::vector<PolymorphicValue> MmaOp::evaluate(
 
   // After removing the broadcast dimensions, the format should be
   // [M, K] x [K, N] compatible with aten::matmul format.
-  return {in_a.matmul(in_b)};
+  auto output = in_a.matmul(in_b);
+
+  // ATen preserves the input dtype whereas MmaOP generates float outputs.
+  // Cast to the dtype of the MmaOp output for consistency.
+  if (tv_a->getDataType() != out()->getDataType().value()) {
+    output = output.to(data_type_to_aten(out()->getDataType().value()));
```

Comment:
Note that this is still a wrong implementation. I believe `at::matmul (bfloat->bfloat) ==> at::cast(bfloat->float)` is less precise than `MmaOp (bfloat->float)`. `MmaOp (bfloat->float)` accumulates in float and never casts it back to the lower precision `bfloat`. However, `at::matmul ==> at::cast` (as in this implementation) introduces an **extra** float->bfloat->float roundtrip and therefore loses some precision. 

I'm fine with this code change as is because it's a strict improvement -- MmaOp::evaluate didn't even get the output type right. However, I'd put this warning in code comment and fix it soonish.

Reply:
Right. There will some precision lost. Will add a comment.
Since `at::matmul` does not allow HH->S matmuls, the nvFuser matmul API will also most likely not support it (@protonu is this assumption valid?).

Once the evaluate method here is updated to handle `MmaOp + castOp` (which is what we should get from the frontend), evaluate can simply return the original output and without any such round trip casting.